### PR TITLE
Replace Slice optional inputs removal to adjustment

### PR DESCRIPTION
**Pull request by dkurt**
### Pull Request Readiness Checklist

resolves https://github.com/opencv/opencv/issues/24671

related discussion: https://github.com/opencv/opencv/pull/24655#discussion_r1416744142

TODO:
- [x] Test with attention layer PR for which initial subgraph has been added.

<details>
<summary>patch</summary>

```patch
--- a/modules/dnn/src/onnx/onnx_graph_simplifier.cpp
+++ b/modules/dnn/src/onnx/onnx_graph_simplifier.cpp
@@ -313,18 +313,30 @@ class AttentionSubGraph : public Subgraph {
         att_add = addNodeToMatch("Add", addNodeToMatch(""), att_matmul);
 
         // v_path
-        slice_v = addNodeToMatch("Slice", att_add, addNodeToMatch(""), addNodeToMatch(""));
+        slice_v = addNodeToMatch("Slice", {att_add,
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch("")});
         int reshape_v = addNodeToMatch("Reshape", slice_v, addNodeToMatch(""));
         int transpose_v = addNodeToMatch("Transpose", reshape_v);
 
         // q_path
-        slice_q = addNodeToMatch("Slice", att_add, addNodeToMatch(""), addNodeToMatch(""));
+        slice_q = addNodeToMatch("Slice", {att_add,
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch("")});
         reshape_q = addNodeToMatch("Reshape", slice_q, addNodeToMatch(""));
         int transpose_q = addNodeToMatch("Transpose", reshape_q);
         div_q = addNodeToMatch("Div", transpose_q, addNodeToMatch(""));
 
         // k_path
-        slice_k = addNodeToMatch("Slice", att_add, addNodeToMatch(""), addNodeToMatch(""));
+        slice_k = addNodeToMatch("Slice", {att_add,
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch("")});
         int reshape_k = addNodeToMatch("Reshape", slice_k, addNodeToMatch(""));
         int transpose_k = addNodeToMatch("Transpose", reshape_k);
 
@@ -422,16 +434,28 @@ class AttentionSingleHeadSubGraph : public Subgraph {
         att_add = addNodeToMatch("Add", addNodeToMatch(""), att_matmul);
 
         // v_path
-        slice_v = addNodeToMatch("Slice", att_add, addNodeToMatch(""), addNodeToMatch(""));
+        slice_v = addNodeToMatch("Slice", {att_add,
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch("")});
         int transpose_v = addNodeToMatch("Transpose", slice_v);
 
         // q_path
-        slice_q = addNodeToMatch("Slice", att_add, addNodeToMatch(""), addNodeToMatch(""));
+        slice_q = addNodeToMatch("Slice", {att_add,
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch("")});
         int transpose_q = addNodeToMatch("Transpose", slice_q);
         div_q = addNodeToMatch("Div", transpose_q, addNodeToMatch(""));
 
         // k_path
-        slice_k = addNodeToMatch("Slice", att_add, addNodeToMatch(""), addNodeToMatch(""));
+        slice_k = addNodeToMatch("Slice", {att_add,
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch(""),
+                                             addNodeToMatch("")});
         int transpose_k = addNodeToMatch("Transpose", slice_k);
 
         // qk
```

</details>

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 1]
```diff
@@ -200,35 +200,17 @@ class RemoveSliceAllOptionalInputsSubgraph : public Subgraph {
...
-                }
-            }
-            return true;
+    virtual void finalize(const Ptr<ImportGraphWrapper>&,
+                          const Ptr<ImportNodeWrapper>& fusedNode,
+                          std::vector<Ptr<ImportNodeWrapper> >&) CV_OVERRIDE
+    {
+        opencv_onnx::NodeProto* node = fusedNode.dynamicCast<ONNXNodeWrapper>()->node;
+        for (int i = num_inputs_; i < 5; ++i) {
+            node->add_input("");
```
Comment (fengyuentau) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: Doesn't it lose connections to the existing constants (starts, ends, axes and steps)?

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: No, It should append missed connections to `axes` and `steps` pointing to a fake initializer added by https://github.com/opencv/opencv/blob/850ebec135bb7904de4a1b9abbbf32b9937c1c06/modules/dnn/src/onnx/onnx_graph_simplifier.cpp#L71

Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: How does this appending happen? What I see here is input names are missed already.

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: For example, `node` which found with 3 real inputs (data, starts, ends) appends by up to 5 inputs with `add_input`:
```
inputs: "data_name", "starts_name", "ends_name", "", ""
```

`net.add_initializer();` adds an empty tensor with an empty name. So the final version of the Slice refers to it's tensor.

I have observed such behavior with `resize_nearest_unfused_opset11_torch1.4` model:
image

Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: Yes, I got you point here. But actually my problem is why this

```.diff
- setFusedNode("Slice", std::vector<int>{input, starts, ends});
+ setFusedNode("Slice", inputs);
```

does not drop the `starts` and `ends` constants.

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: They were added before. Just there is no overload with 5 inputs and vector is used:
```cpp
std::vector<int> inputs{input, starts, ends};
for (size_t i = 3; i < num_inputs_; i++) { // axes and steps
    inputs.push_back(addNodeToMatch(""));
}
```

Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: Ahhh, sorry, I missed that. I thought it was `input`.

Comment (Abdurrahheem) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: @dkurt is there a way for me to check these fixes locally before merging?

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: @Abdurrahheem, do the following:
```
git remote add dkurt https://github.com/dkurt/opencv
git fetch dkurt adjust_slice_optional_inputs
git checkout dkurt/adjust_slice_optional_inputs
```

Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: > @dkurt is there a way for me to check these fixes locally before merging? Wanna make sure if yolox test do pass.

git cherry-pick should also work right? The simplest way is do the same changes as this PR manually, so you don’t need to mess around with git operations.

**Review by fengyuentau — APPROVED** [review]

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp (hunk 1 above)
Comment (Abdurrahheem) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:212]: @dkurt just checked test with the new patch. Everything passes related to yolo detectors. Thank you for a quick fix!

**Pull request merged**